### ISSUE TITLE
"Cannot read property '$options' of undefined" when saving and manipulating data in watchers

### ISSUE THREAD

**Issue by danielelkington** [vuejs/vue-test-utils#1130]
### Version
1.0.0-beta.29

### Reproduction link
[https://codesandbox.io/s/20yl7w8y6y](https://codesandbox.io/s/20yl7w8y6y)

### Steps to reproduce
In the Code Sandbox run the tests - they will fail.
Due to the way Code Sandbox runs tests if you re-run them they'll pass the second time. You can make them run properly again by refreshing the page or deleting a semicolon somewhere in the file.

Commenting out either line in the `handler` function will cause the tests to pass.

### What is expected?
Inside watchers, you should be able to save the component data into a variable and then set it.

### What is actually happening?
Saving component data in a variable and then setting the data inside a watcher fails tests in the latest version of vue/vue-test-utils. This has caused many tests to suddenly fail in our application.

Stack trace (from CodeSandbox):
```
Cannot read property '$options' of undefined

TypeError: Cannot read property '$options' of undefined
    at updateChildComponent (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:2984:8)
    at prepatch (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:4388:5)
    at patchVnode (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:6180:7)
    at VueComponent.patch [as __patch__] (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:6352:9)
    at VueComponent.Vue._update (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:2815:19)
    at VueComponent.updateComponent (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:2933:10)
    at Watcher.get (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3330:25)
    at Watcher.run (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3405:22)
    at flushSchedulerQueue (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3163:13)
    at queueWatcher (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3252:9)
    at Watcher.update (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3395:5)
    at Dep.notify (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:740:13)
    at Object.reactiveSetter [as myData] (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:1065:11)
    at VueComponent.proxySetter [as myData] (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3482:26)
    at VueComponent.handler (https://20yl7w8y6y.codesandbox.io/test.spec.js:21:21)
    at VueComponent.Vue.$watch (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3797:12)
    at createWatcher (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3754:13)
    at initWatch (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3736:7)
    at initState (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3499:5)
    at VueComponent.Vue._init (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:4877:5)
    at new VueComponent (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:5024:12)
    at createComponentInstanceForVnode (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:4546:10)
    at init (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:4377:45)
    at createComponent (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:5850:9)
    at createElm (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:5797:9)
    at VueComponent.patch [as __patch__] (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:6347:7)
    at VueComponent.Vue._update (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:2812:19)
    at VueComponent.updateComponent (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:2933:10)
    at Watcher.get (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3330:25)
    at new Watcher (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:3319:12)
    at mountComponent (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:2940:3)
    at VueComponent.Vue.$mount (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:8898:10)
    at VueComponent.Vue.$mount (https://20yl7w8y6y.codesandbox.io/node_modules/vue/dist/vue.common.dev.js:11692:16)
    at mount (https://20yl7w8y6y.codesandbox.io/node_modules/@vue/test-utils/dist/vue-test-utils.js:8649:21)
    at shallowMount (https://20yl7w8y6y.codesandbox.io/node_modules/@vue/test-utils/dist/vue-test-utils.js:8677:10)
    at Object.eval (https://20yl7w8y6y.codesandbox.io/test.spec.js:34:33)
    at https://codesandbox.io/static/js/vendors~sandbox.5596247f.chunk.js:1:180690
    at new Promise (<anonymous>)
    at t.callAsyncFn (https://codesandbox.io/static/js/vendors~sandbox.5596247f.chunk.js:1:180353)
    at https://codesandbox.io/static/js/sandbox.37038f94.js:1:186582
    at g (https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:11215)
    at Generator._invoke (https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:11003)
    at Generator.e.(anonymous function) [as next] (https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:11394)
    at r (https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:702)
    at u (https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:912)
    at https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:971
    at new Promise (<anonymous>)
    at https://codesandbox.io/static/js/common-sandbox.458a22cf.chunk.js:1:853
    at https://codesandbox.io/static/js/sandbox.37038f94.js:1:186850
    at https://codesandbox.io/static/js/sandbox.37038f94.js:1:185325
```

**Comment by eddyerburgh** [comment]
This is caused by the `sync` mode change in beta.29.

The workaround is to set `sync` to `false`:

```js
shallowMount(component, { sync: true })
```

We may need to change the Vue Test Utils API and remove synchronous rendering entirely, which would require explicit calls to a `flushPromises` helper function after each change that triggers a watcher.

**Comment by danielelkington** [comment]
Thanks @eddyerburgh for the quick response.

I think this is quite a serious problem - it means many people who use watchers will suddenly start seeing tests fail with this error message - and confusingly it only seems to happen when running at least two tests - the first test always seems to work no problems. When our team tried to upgrade to beta.29 today hundreds of tests began failing (we've had to downgrade back to beta.28). Assuming no proper fix for this is possible we'll need to add hundreds of flushPromises() calls throughout our tests.

Will the docs be updated soon with advice about this? And is no fix where the existing API is maintained possible?

**Comment by eddyerburgh** [comment]
@danielelkington Yes it's a very serious problem—bugs with sync mode is the reason we're still in beta.

The only solution to avoid any edge cases and maintain a synchronous API in Vue Test Utils is to maintain a sync version of Vue and fix issues like this in the core repo, but I don't think that's realistic. 

On top of that, Vue 3 probably won't include synchronous updating either.

If we decide to remove sync mode, I'll update the docs, add warnings, and make the migration as smooth as possible. But I understand it's a huge change that will affect a lot of codebases, and I'm sorry for the extra work it will cause.

**Comment by AtofStryker** [comment]
@danielelkington  and @eddyerburgh I thin we have a similar problem as well. After upgrading to beta-29, some tests fail with:

Cannot read property '_transitionClasses' of undefined
Or
Cannot read property '$scopedSlots' of undefined

After investigating a bit further, it looks to be related to the sync/flushing of the current or previous test. Setting sync to false suppresses the error

Not sure if I should create a separate issue since the two seem related. I've been trying to reproduce the error in codesandbox to no avail.

**Comment by eddyerburgh** [comment]
@AtofStryker No need to create an extra issue, since the root cause is the same.

We've decided to remove sync mode, I'll write-up the reasons why in the next few days, and we'll remove sync mode from the next Vue Test Utils version.

**Comment by VladZen** [comment]
Seems to be a good news: i have about 400 tests, using sync mode and now it's gonna be removed, nice work! You know, i appreciate all you're doing for this project, but don't you think it's unfair to implement such a good feature and after users make tests for their projects you say like "It's very buggy, it will be removed" instead of "Okay, we have bugs, but let's try to fix them"?

**Comment by danielelkington** [comment]
@VladZen We have over 1500 tests using sync mode so I feel your pain, but I also respect that vue-test-utils is still in beta, so it's not like a promise was ever made that the API would remain stable. Using the library to write a lot of tests was thus always a bit of a risk. I trust that @eddyerburgh and the core team are making decisions that will make Vue and vue-test-utils the best that they can be long-term, so I'm sure this will cause some short-term pain in exchange for a stable and quality library in the long term.

**Comment by eddyerburgh** [comment]
@VladZen we tried to fix them with multiple approaches, and all of them led to different bugs. This is the best approach for the long-term.

Thank you @danielelkington

**Comment by sobolevn** [comment]
So, how can this be fixed?

```ts
  test('should not pass validation with invalid inputs', async () => {
    expect.hasAssertions()

    // make `city` and `address` empty
    const locationInvalid = deepcopy(location)
    locationInvalid.city = ''
    locationInvalid.address = ''

    await flushPromises()

    const wrapper = mount(TLocations, {
      'sync': false,
      'attachToDocument': true,
      localVue,
      'propsData': {
        'location': locationInvalid,
      },
    })

    await flushPromises()

    // assert validation result
    const formResult = await wrapper.vm.$validator.validateAll()

    await flushPromises()
    expect(formResult).toBe(false)
  })
```

Gives me an error:

```
Summary of all failing tests
 FAIL  tests/components/templates/TLocations.spec.ts (10.269s)
  ● unit tests for TLocations component › should not pass validation with invalid inputs

    TypeError: Cannot read property '$scopedSlots' of undefined

      at updateChildComponent (node_modules/vue/dist/vue.runtime.common.dev.js:4091:27)
      at prepatch (node_modules/vue/dist/vue.runtime.common.dev.js:3119:5)
      at patchVnode (node_modules/vue/dist/vue.runtime.common.dev.js:6282:7)
      at updateChildren (node_modules/vue/dist/vue.runtime.common.dev.js:6167:9)
      at patchVnode (node_modules/vue/dist/vue.runtime.common.dev.js:6293:29)
      at updateChildren (node_modules/vue/dist/vue.runtime.common.dev.js:6167:9)
      at patchVnode (node_modules/vue/dist/vue.runtime.common.dev.js:6293:29)
```

I am using `"@vue/test-utils": "^1.0.0-beta.29",`

**Comment by cleverboy32** [comment]
@sobolevn have you resoved ? and how ?

**Comment by sobolevn** [comment]
Nope, I have just ignored it for now. Hopefully I can ignore this case.
But, this is 100% required to be fixed.

**Comment by sobolevn** [comment]
Now I cannot ignore it anymore. @eddyerburgh what should I do?

**Comment by danielelkington** [comment]
@sobolevn When we had this issue we downgraded to 1.0.0-beta.28 where this issue doesn't seem to occur. We're waiting for the next release when sync mode will be removed and there will be some sort of migration guide - we plan to migrate our tests out of sync mode then.

**Comment by sobolevn** [comment]
It solves this issue, but I now have new ones. In other words: nope, it does not work for me.

Anyway, @danielelkington, thanks for the suggestion! 👍

**Comment by cleverboy32** [comment]
@danielelkington  its works for me and thanks

**Comment by fsblemos** [comment]
@sobolevn `sync: false` solved for me https://github.com/vuejs/vue-test-utils/issues/1130#issuecomment-460558486

**Comment by georgwittberger** [comment]
I was facing a similar error message recently when I used the `pagination.sync` prop on a [Vuetify Data Table](https://vuetifyjs.com/en/components/data-tables#external-sorting):

```
TypeError: Cannot read property '$scopedSlots' of undefined
```

I suppose that the `sync` modifier caused this problem because before I used this special prop of the data table all unit tests ran successfully even with version 1.0.0-beta.29 of Vue Test Utils.

As a workaround I downgraded to 1.0.0-beta.28 again but in order to make the tests for the data table work I had to mount my component using the `sync: false` option, as proposed before. This gave me some more error messages of this kind:

```
Cannot read property '_transitionClasses' of undefined
```

The solution for me was to write the tests completely in async style. Like this:

```javascript
import {mount} from '@vue/test-utils';
import flushPromises from 'flush-promises';
// more imports omitted here

describe('MyComponent.vue', () => {
  it('does something beautiful', async () => {
    const wrapper = mount(MyComponent, {sync: false});
    await flushPromises();
    expect(wrapper.find('.my-component').text()).toContain('Unicorns');
  });
});
```

**Comment by kamok** [comment]
` sync: false` works. Caution that you need it on every `shallowMount` or `mount` in a single test file. I had `shallowMount` without sync, and then `mount` with sync and still threw an error at the mount instead.

Is there a clean way to ensure that every single shalllowMount calls gets passed `sync: false` without adding it manually?

**Comment by doutatsu** [comment]
@kamok Just use a `beforeEach()` and then set new data on top of mounted component (sorry for poor indentation):
```js
    let test;

    beforeEach(() => {
      test = mount(Test, {sync: false });
    });

    it('does a thing', async () => {
      test.setProps({
        tableData: [
          {
            object: {
              string: "test",
            },
            },
          },
        ],
      });

      await flushPromises();

      ...
    });
```

**Comment by acacha** [comment]
@kamok Thanks you comment: "Caution that you need it on every shallowMount or mount in a single test file" helps to me because sync: false  doesn't work to me because I used it only in one mount no in all test mounts/shallowMounts

**Comment by chrisnicola** [comment]
Finding that when running all my mocha tests it appears that any mount call in any test files can trigger this problem. So it looks like I have to go around and add `sync: false` and `await wrapper.vm.$nextTick()` throughout all the tests.

This seems wrong to me. Do I have this right? Is this being addressed in a future version of vue-test-utils?

**Comment by ghost** [comment]
@chrisnicola Same resolution here, waiting on input from the vue-test-utils team to see if there are any better alternatives, will start dropping sync mode from now on.

**Comment by chrisnicola** [comment]
I'm concerned there has not been much activity since this was reported 8 months ago now. Is testing not a major concern for VueJS? From what I can see there is a lot of focus is on getting 3.0 shipped but there still isn't a solid reliable testing option for Vue yet.

I mean I understand this is still in beta, but I also don't believe there are any alternatives. Avoriaz is no longer in active development as I understand it.

Considering this appears to be a regression from the previous beta release, is there any reason that this change wasn't simply reverted? Or why the proposed fix of removing the `sync` option and defaulting to `false` behaviour wasn't implemented?

**Comment by dobromir-hristov** [comment]
`sync` has been removed in version 30. All tests are now `async` by default, which means watchers should work as they do in your app. 
We can close this one I think, what do you say @eddyerburgh ?

**Comment by danielelkington** [comment]
We have upgraded to beta.30, and have found that the process of fixing our ~2000 tests was quite straightforward (just needed to add a lot of `flushPromises()` calls). Watchers now work as expected. Thank you.

**Comment by tmorehouse** [comment]
Perhaps `flushPromises` should be included as a method that can be imported from test utils?

**Comment by andreixk** [comment]
> ... I also respect that vue-test-utils is still in beta, so it's not like a promise was ever made that the API would remain stable. ...

1. it's been in beta since 2017
2. beta means minor changes possible, not removing core functionality
3. if we are to assume API will not remain stable, then why use it at all? by using a library that will change fundamentally in 2 months, then i'm knowingly creating technical debt for everyone involved in the project.

**Comment by Jatin-Kamboj** [comment]
@sobolevn  Thanks this worked for me  `sync: false` solved for me #1130 (comment)
Really Appreciated.
